### ISSUE TITLE
ENH: Skew symmetric/matrix cross product function

### ISSUE THREAD

**Issue by Mark2000** [numpy/numpy#24687]
### Proposed new feature or change:

Dynamicists frequently use the matrix representation of the cross product (https://en.m.wikipedia.org/wiki/Cross_product#Conversion_to_matrix_multiplication), and the lack of a skew symmetric operator in NumPy is a frequent source of annoyance.

I propose either adding a skew function to linalg that returns the skew symmetric matrix of a vector `np.cross(a, np.identity(a.shape[0]) * -1)` (from https://stackoverflow.com/questions/66707295/numpy-cross-product-matrix-function), or more radically, making the second argument `b` of np.cross optional and have `np.cross(a)` return the matrix representation of [a x]. For the latter, would need to consider the interaction with other kwargs and different input shapes.

**Comment by WarrenWeckesser** [comment]
This seems a bit too specialized for NumPy.  The SciPy module [`scipy.linalg.special_matrices`](https://docs.scipy.org/doc/scipy/reference/linalg.html#special-matrices) might be a better fit.

**Comment by miccoli** [comment]
Some very quick comments.
- I would not consider `np.skew` more specialized than, say, `np.kron`.
- `np.cross(a, np.identity(a.shape[0]) * -1)` works only in $R^3$, but `np.skew(a)` should/could be generalized in $R^n$.
- `np.cross(a)` is not a good idea, in my opinion, because it would be very difficult to generalize for $n \neq 3$.
- what about implementing the Levi-Civita symbol in `np.einsum`?

**Comment by Mark2000** [comment]
As far as I know, the operator can’t be meaningfully generalized into n dimensions as it is only well defined for dimensions the size of an upper triangle/those where a cross product exists (i.e. 1, 3, 7, 15, etc.). Please correct me if there’s a good way of generalizing to other dimensions (maybe L-C does?  could find that one way or another). As a counterargument, I’d say that like np.cross, this is probably only commonly useful in n=3 and generalizing might not be worth it in this case; perhaps there are uses for other dimensions that I’m not familiar with. 

As for calling it `skew`, that seems to follow a convention that exists elsewhere and is probably the best name; however, to play devils advocate, there are many other ways of generating a skew symmetric matrix from a vector - it’s just that few of them have any common use other than this one. 

I have no experience with einsum but read about the Levi-Civita symbol when researching this; what would the notation look like in einsum?

**Comment by miccoli** [comment]
> As far as I know, the operator can’t be meaningfully generalized into n dimensions as it is only well defined for dimensions the size of an upper triangle/those where a cross product exists (i.e. 1, 3, 7, 15, etc.).

Yes, `skew(a)` should only be defined when the last dimension of `a` has length $\frac12 n(n-1)$, so $1,3,6,10,15,\ldots$
In other terms for $R^n$ the expression `skew(a) @ b` is meaningful only if $a$ is a is bivector ($\frac12 n(n-1)$ components) and $b$ a vector ($n$ components).

> I have no experience with einsum but read about the Levi-Civita symbol when researching this; what would the notation look like in einsum?

I've not worked out this point, but I suspect that `np.skew(a) @ b` could be written efficently using `np.einsum` and some smart trick... but probably this path is really too specialized, with only marginal improvements.

**Comment by charris** [comment]
> As far as I know, the operator can’t be meaningfully generalized into n dimensions

It can be. An ordered list of `n - 1` vectors defines a parallelepiped  with oriented volume V. pick an oriented vector at right angles to the subspace determined by those vectors and give it a length equal to V. You need to pick an orientation for this to work. Related to the Hodge star of `n - 1` forms :)  Agree that it isn't a binary operator.

**Comment by Mark2000** [comment]
@miccoli My subconscious really wanted the 7d cross product matrix to just be a skew symmetric matrix constructed with 7 elements in the upper triangle, despite that clearly not being possible. 

It seems like there are two (potentially linked) products that could come from this discussion:
- A matrix cross function `np.cross(a)`
- A generalized $\frac{1}{2}n(n-1)$ (or even $n$) dimensional `np.skew(a)` function

where the former is happens to be a special case of the latter for `a` with a last dimension shape 3, which seems appropriate given that `np.cross` is already (arbitrarily? pragmatically?) limited to 3 dimensional crosses. I know the former is practically useful for me and others based on conversations with colleagues and the existence of multiple stack overflow threads; is the latter actually a feature desired by anyone?

**Comment by miccoli** [comment]
> I know the former is practically useful for me and others based on conversations with colleagues and the existence of multiple stack overflow threads; is the latter actually a feature desired by anyone?

Fair observation: for sure most applications would be for $n=3$, while $n\neq 3$ would be in the realm of the “too specialized” cases ... much better to avoid a generalization that leads to consistent but rarely used cases.

**Comment by rgommers** [comment]
> This seems a bit too specialized for NumPy. The SciPy module [`scipy.linalg.special_matrices`](https://docs.scipy.org/doc/scipy/reference/linalg.html#special-matrices) might be a better fit.

I agree with @WarrenWeckesser here. The current discussion seems useful and interesting, so by all means continue it - but aiming for inclusion in SciPy seems better. We don't really want to expand `numpy.linalg`, SciPy contains a superset.

**Comment by miccoli** [comment]
> > This seems a bit too specialized for NumPy. The SciPy module [`scipy.linalg.special_matrices`](https://docs.scipy.org/doc/scipy/reference/linalg.html#special-matrices) might be a better fit.

So maybe `a = scipy.linalg._special_matrices.cross(c)` with `c` of shape `(N*(N-1)/2,)` and `a` `(N,N)`?
BTW `scipy.linalg.cross` is free...

My only doubt here is that while numpy allows for *stacking* matrices/vectors, `scipy.linalg._special_matrices` typically doesn't (the first statement is almost always `c = np.asarray(c).ravel()`).

Since writing the desired function in pure python is not that difficult, I was assuming that inclusion in numpy was also in order to have a fast and efficient implementation.

**Comment by rgommers** [comment]
Support for stacking is desirable for `scipy.linalg` as well. That most functions don't support that now is a combination of history and lack of bandwidth to implement it. I think you can keep stacking support for a new scipy function.

**Comment by Mark2000** [comment]
Seems like the consensus is leaning towards scipy. Will look into opening a PR over there when I get a chance.
